brave-core/src/main/java/com/github/kristofa/brave/InheritableServerClientAndLocalSpanState.java

```diff
@@ -0,0 +1,137 @@
+package com.github.kristofa.brave;
+
+import java.util.Deque;
+import java.util.concurrent.LinkedBlockingDeque;
+
+import com.github.kristofa.brave.internal.Util;
+import com.twitter.zipkin.gen.Endpoint;
+import com.twitter.zipkin.gen.Span;
+
+/**
+ * {@link ServerClientAndLocalSpanState} implementation that keeps trace state using {@link InheritableThreadLocal}
+ * variables and provides local span inheritence from parent to children.
+ * <p>
+ * Important note: when using {@link InheritableServerClientAndLocalSpanState}, tracers must
+ * {@link LocalTracer#finishSpan() finish spans} or clear the local span at
+ * completion of the local trace span to avoid linking spans with incorrect
+ * parents and avoid leaking spans and associated memory.
+ */
+public final class InheritableServerClientAndLocalSpanState implements ServerClientAndLocalSpanState, InheritableState {
+
+    private final InheritableThreadLocal<ServerSpan> currentServerSpan =
+            new InheritableThreadLocal<ServerSpan>() {
+                @Override
+                protected ServerSpan initialValue() {
+                    return ServerSpan.EMPTY;
+                }
+            };
+
+    private final InheritableThreadLocal<Span> currentClientSpan = new InheritableThreadLocal<Span>();
+
+    private final InheritableThreadLocal<Deque<Span>> currentLocalSpan =
+            new InheritableThreadLocal<Deque<Span>>() {
+                @Override
+                protected Deque<Span> initialValue() {
+                    return new LinkedBlockingDeque<Span>();
+                }
+            };
+
+    private final Endpoint endpoint;
+
+    /**
+     * Constructor
+     *
+     * @param ip Int representation of ipv4 address.
+     * @param port port on which current process is listening.
+     * @param serviceName Name of the local service being traced. Should be lowercase and not <code>null</code> or empty.
+     */
+    public InheritableServerClientAndLocalSpanState(int ip, int port, String serviceName) {
```

Comment:
switch to Endpoint instead. That way people can use ipv6 etc and we don't need to copy/paste a bunch of things